examples/authentication/client/src/main.rs

```diff
@@ -19,54 +19,70 @@
 //! This example uses the Google Identity Platform.
 //! https://developers.google.com/identity/
 
-use futures_util::future;
-use hyper::{service::Service, Body, Request, Response, Server, StatusCode};
+use hyper::Server;
 use log::{info, warn};
-use std::{
-    collections::HashMap,
-    task::{Context, Poll},
-};
 use structopt::StructOpt;
-use tokio::sync::{
-    mpsc::{self, Sender},
-    oneshot,
-};
-use url::{form_urlencoded, Url};
+use tokio::sync::{mpsc, oneshot};
+
+mod auth_client;
+mod redirect_handler;
 
 #[derive(StructOpt, Clone)]
 #[structopt(about = "OpenID Connect Client Example.")]
 pub struct Opt {
     #[structopt(
         long,
         help = "Address of the Oak application to connect to.",
-        default_value = "127.0.0.1:8080"
+        default_value = "https://127.0.0.1:8080"
     )]
     address: String,
+    #[structopt(
+        long,
+        help = "Address of the authentication server.",
+        default_value = "https://localhost:8081"
+    )]
+    auth_server: String,
     #[structopt(
         long,
         help = "Address to listen on for the OAuth redirect.",
         default_value = "127.0.0.1:8089"
     )]
     redirect_address: String,
-    #[structopt(long, help = "Path to the PEM-encoded CA root certificate.")]
-    ca_cert: Option<String>,
-    // TODO(#886): Retrieve client_id from server, rather than flag.
-    #[structopt(long, help = "The OAuth client ID.")]
-    client_id: String,
+    #[structopt(
+        long,
+        help = "Path to the PEM-encoded CA root certificate.",
+        default_value = "examples/certs/local/ca.pem"
+    )]
+    ca_cert: String,
 }
 
 fn main() -> Result<(), Box<dyn std::error::Error>> {
     env_logger::init();
     let opt = Opt::from_args();
-    let redirect_address = opt.redirect_address.parse()?;
-    info!("Listening for redirect on {}", &redirect_address);
+    let mut runner = tokio::runtime::Runtime::new().unwrap();
+    let request_url = runner.block_on(auth_client::get_authentication_request_url(
+        &opt.ca_cert,
+        &opt.auth_server,
+        &opt.redirect_address,
+    ))?;
+    let code = runner.block_on(get_authentication_code(&opt.redirect_address, &request_url))?;
+    info!("Received code: {}", code);
+    Ok(())
+}
 
+/// Gets the OIDC authentication code by opening the default browser and extracting it from the
```

Comment:
Does the browser open in the headless mode?
Also, why do we need a browser instead a single HTTP request (does it require user interaction)?

Reply:
It opens a normal browser window. It requires user interaction: the user has to log in to their Google account or select which one to use if they are already logged in.